sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images/ZE_experimental_bindless_images.asciidoc

```diff
@@ -0,0 +1,577 @@
+    0,                            /* flags, set to read only */
+    ZE_IMAGE_TYPE_2D,             /* type */
+    imageFormat,                  /* format */
+    imageWidth,                   /* width */
+    imageHeight,                  /* height */
+    0,                            /* depth */
+    0,                            /* arrayLevels */
+    0                             /* miplevels */
+};
+
+// Allocate USM for images and return the pointer
+ze_image_usm_alloc_exp_desc_t allocDesc;
+zeMemAllocPitchedExp(hContext, &imageDesc, &allocDesc);
+
+// Attach allocation to the image descriptor
+imageDesc.pNext = &allocDesc;
+
+// Declare the copy region for copying
+ze_copy_region_t copyRegion = {
+  0,                          /* originX */
+  0,                          /* originY */
+  0,                          /* originZ */
+  imageWidth * sizeof(float), /* width */
+  imageHeight,                /* height */
+  0                           /* depth */
+};
+
+// Copy from host to device
+zeCommandListAppendMemoryCopyRegion(hCommandList,
+                                    allocDesc.pMemAlloc /* dstptr */,
+                                    &copyRegion /* dstRegion */,
+                                    pitchDesc.rowPitch /* dstPitch */,
+                                    0 /* dstSlicePitch */,
+                                    imageDataHost.data() /* srcptr */,
+                                    &copyRegion /* srcRegion */,
+                                    imageWidth * sizeof(float) /* srcPitch */,
+                                    0 /* srcSlicePitch */,
+                                    nullptr /* hSignalEvent */,
+                                    0 /* numWaitEvents */,
+                                    nullptr /* phWaitEvents */);
+
+// Create image from memory allocated above
+ze_image_handle_t hImage;
+zeImageCreate(hContext, hDevice, &imageDesc, &hImage);
+
+// Now we can operate on the image, passing `hImage` as an argument
+// to a kernel
+
+// Copy from device to host
+zeCommandListAppendMemoryCopyRegion(hCommandList,
+                                    imageDataHost.data() /* dstptr */,
+                                    &copyRegion /* dstRegion */,
+                                    imageWidth * sizeof(float) /* dstPitch */,
+                                    0 /* dstSlicePitch */,
+                                    allocDesc.pMemAlloc /* srcptr */,
+                                    &copyRegion /* srcRegion */,
+                                    pitchDesc.rowPitch /* srcPitch */,
+                                    0 /* srcSlicePitch */,
+                                    nullptr /* hSignalEvent */,
+                                    0 /* numWaitEvents */,
+                                    nullptr /* phWaitEvents */);
+
+// Once all operations on the image are complete we need to free the memory and
+// destroy the handle
+
+// Free image memory
```

Comment:
memory cannot be freed before image is destroyed - as image is using this memory as a backing stroage,

the order of freeing should be reverse to allocation order:

zeImageDestroy(hImage);
zeMemFree(hContext, allocDesc.pMemAlloc);

Reply:
@HoppeMateusz that's right - this was overlooked in this proposal. You'll see that in the DPC++ extension we destroy image handles and then free the memory - reverse allocation order as you've mentioned. It is meant for this to work in the same order.

Apologies for the confusion here.